### PR TITLE
Add counter for multiple occurences

### PR DESCRIPTION
**Pull request by CanRau**
Closes https://github.com/sindresorhus/slugify/issues/37

That's actually what I already made for [@gaiama/slugger](https://github.com/GaiAma/Coding4GaiAma/tree/master/packages/gaiama-slugger)

only issue so far is, it's not doing the by you preferred `-2-2` 😢 

```js
t.is(countableSlugify('foo-2'), 'foo-2-1'); // or foo-2-2
```
and I'm currently not sure how to properly do it without a rewrite.
I'm open to other implementations or preferences.

`index.d.ts` is probably not correct? Still very new to TS and external declarations.

Also I'm not sure if the used regex `/(?:-\d+?)+?$/` is [ReDoS](https://en.wikipedia.org/wiki/ReDoS) safe or how much that matters here 🤷‍♂

### CODE REVIEW

**Comment by sindresorhus** [comment]
> It cannot be the main slugify() method though as modifying it affects the whole process, not just your code. I'm thinking const countableSlugify = slugify.counter() to create a counter.

The counter cannot be global. It needs to create a new instance.

**Comment by CanRau** [comment]
Not sure I understand, the current implementation is used like so
```js
const countableSlugify = slugify.counter();
t.is(countableSlugify('foo bar'), 'foo-bar');
```

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/slugify/pull/46/files#diff-168726dbe96b3ce427e7fedce31bb0bcR71 is global. If `slugify` is used in multiple places in the dependency tree, they will share this, which will cause really weird problems.

**Comment by CanRau** [comment]
You're right 😅 
That should fix it, I've added tests for it as well.
Is that how you "envisioned" it?

File: index.d.ts [hunk 1]
```diff
@@ -141,3 +141,10 @@ declare function slugify(
 ): string;
 
 export = slugify;
+
+declare function counter(
+	string: string,
+	options?: slugify.Options
+): string;
+
+declare function reset(): void;
```
Comment (sindresorhus) [inline index.d.ts]: The TS types are no longer correct.

**Comment by sindresorhus** [comment]
Yes, perfect. That's exactly what I had in mind.

**Comment by sindresorhus** [comment]
Can you add docs (readme + TypeScript file)?

**Comment by CanRau** [comment]
I updated everything as good as I could.
Please let me know if that works for you.

**Comment by CanRau** [comment]
Should the `countableSlugify` count empty strings like
```js
countableSlugify(""); // => ""
countableSlugify(""); // => "-2"
```
or should it always return `""` ? 🤔

File: index.d.ts [hunk 2]
```diff
@@ -133,11 +133,32 @@ slugify('fooBar 123 $#%');
...
-	string: string,
-	options?: slugify.Options
-): string;
+declare const slugify: {
+	(
+		string: string,
+		options?: slugify.Options
+	): string;
+
+	counter: () => {
```
Comment (sindresorhus) [inline index.d.ts:153]: This needs a doc comment.

File: index.d.ts [hunk 3]
```diff
@@ -133,11 +133,32 @@ slugify('fooBar 123 $#%');
...
+		(
+			string: string,
+			options?: slugify.Options
+		): string;
+
+		reset(): void;
```
Comment (sindresorhus) [inline index.d.ts:159]: This needs a doc comment.

File: readme.md [hunk 4]
```diff
@@ -164,6 +164,45 @@ slugify('_foo_bar', {preserveLeadingUnderscore: true});
...
+## Section 1
+
+### Example
+
+## Section 2
+
+### Example
+```
+
+You can then use `slugify.counter()` to generate unique HTML `id`'s to ensure anchors will link to the right headline.
```
Comment (sindresorhus) [inline readme.md:204]: index.d.ts should be in sync with the readme docs.

File: test.js [hunk 5]
```diff
@@ -131,3 +131,36 @@ test('leading underscore', t => {
...
+
+	t.is(countableSlugify('foo'), 'foo');
+	t.is(countableSlugify('foo'), 'foo-2');
+	t.is(countableSlugify('foo 1'), 'foo-1');
+	t.is(countableSlugify('foo-1'), 'foo-1-2');
+	t.is(countableSlugify('foo-1'), 'foo-1-3');
+	t.is(countableSlugify('foo'), 'foo-3');
+	t.is(countableSlugify('foo'), 'foo-4');
+	t.is(countableSlugify('foo-1'), 'foo-1-4');
+	t.is(countableSlugify('foo-2'), 'foo-2-1'); // Or foo-2-2 ??
```
Comment (sindresorhus) [inline test.js:150]: This comment should be removed.

**Comment by sindresorhus** [comment]
> or should it always return "" ? 🤔

This. Add tests for that.

File: index.d.ts (hunk 2 above)
Comment (CanRau) [reply index.d.ts:166]: like so? synched to readme example.

File: index.d.ts (hunk 3 above)
Comment (CanRau) [reply index.d.ts:192]: Like so, or is it overkill?

**Comment by sindresorhus** [comment]
Make sure `$ npm test` passes locally.

**Comment by CanRau** [comment]
Thanks for the reminder, I only ran `npx ava --watch` yesterday and forgot about `xo` & `tsd` 😬

File: index.d.ts [hunk 6]
```diff
@@ -113,31 +113,84 @@ declare namespace slugify {
...
+	//=> 'ya-lyublyu-edinorogov'
+	```
+	*/
+	(
+		string: string,
+		options?: slugify.Options
+	): string;
+
+	/**
+	Creates a counter to ensure uniqueness.
```
Comment (sindresorhus) [inline index.d.ts:145]: This and the readme description should be in sync. The description is different in the readme. Also missing the `#### Use case example of counter` part.

File: readme.md [hunk 7]
```diff
@@ -164,6 +164,45 @@ slugify('_foo_bar', {preserveLeadingUnderscore: true});
 //=> '_foo-bar'
 ```
 
+### slugify.counter()
```
Comment (sindresorhus) [inline readme.md:167]: The `.reset()` method needs to be documented in the readme too.

**Comment by sindresorhus** [comment]
Looks great now. Thanks for working on this 👍

**Comment by CanRau** [comment]
Thanks a lot for your patience & guidance @sindresorhus very much appreciated 🙏 
I learned a lot 🎉

**Pull request merged**